### ISSUE TITLE
Is Kokkos::Complex trivially copyable?

### ISSUE THREAD

**Issue by kddevin** [kokkos/kokkos#2577]
In 
https://github.com/trilinos/Trilinos/issues/6295#issuecomment-558323602 ,
gcc 8.3 is giving a compilation warning-as-error when memcpy is used to copy Kokkos::Complex.  

Is it possible to make Kokkos::Complex trivially copyable?

Trilinos is trying to spin up gcc 8.3 PR builds, so resolving this issue, either in Tpetra or Kokkos, is necessary.

Thanks.

**Comment by crtrott** [comment]
I have no clue why it wouldn't be.

**Comment by dhollman** [comment]
This may be related to the over-alignment (e.g., because the `memcpy` may be unsafe unless you check the alignment of the destination), but AFAICT from the standard that doesn't actually affect *TriviallyCopyable* requirements.

**Comment by dhollman** [comment]
Is this a bug on develop or master?

**Comment by crtrott** [comment]
Fails on develop too:
```
#include<Kokkos_Core.hpp>
#include<cstdio>
int main(int argc, char* argv[]) {

  if(std::is_trivially_copyable<Kokkos::complex<double>>::value)
    printf("Yeah\n");
  else
    printf("Duh\n");
}
```

**Comment by masterleinad** [comment]
The `volatile` overloads seem to be responsible: https://godbolt.org/z/TxWtjm.

**Comment by masterleinad** [comment]
More precisely, the problem is that the volatile copy assignments are not trivial and can't be.

**Comment by kddevin** [comment]
Is this feature something that can be fixed in Kokkos?  
Or do we have to workaround it in Trilinos (avoiding memcpy for scalar_t = Kokkos::Complex)?

In @masterleinad 's example, is the overloaded operator= needed?  Or can a default copy constructor be used?  E.g., this code works in that website:
```
#include <type_traits>
#include <complex>
#include <iostream>

template <typename RealType>
struct complex
{
   private:
  RealType re_{};
  RealType im_{};

 public:
  using value_type = RealType;
  complex(RealType r, RealType i) : re_(r), im_(i) {}
  void print() { std::cout << re_ << " " << im_ << std::endl;}
};

int main()
{
  static_assert(std::is_trivially_copyable<complex<double>>::value, "");
  struct complex<int> a(10, 2);
  struct complex<int> b = a;
  a.print();
  b.print();
}
```

**Comment by masterleinad** [comment]
> Is this feature something that can be fixed in Kokkos?
> Or do we have to workaround it in Trilinos (avoiding memcpy for scalar_t = Kokkos::Complex)?

This will not be addressed for the 3.0 release since it is pre-existent behavior and (likely) non-trivial to fix. 

> In @masterleinad 's example, is the overloaded operator= needed? Or can a default copy constructor be used? E.g., this code works in that website:

Yes, the overloaded copy assignment operator is (currently) necessary for `atomic` operations.

**Comment by kddevin** [comment]
@crtrott @mhoemmen 
Is Kokkos::Complex still needed in Trilinos?   
@masterleinad says it will not be fixed in Kokkos.
If we still need it, we'll have to write copy loops instead of using memcpy.

**Comment by mhoemmen** [comment]
@kddevin wrote:
> Is `Kokkos::complex` still needed in Trilinos?

Yes.  Salinas needs `Scalar=complex<double>`, and Salinas needs the CUDA build to work.  `std::complex<T>` does not work in CUDA land.  This is why we use `Kokkos::complex<T>`.





> If we still need it, we'll have to write copy loops instead of using memcpy.

As long as `sizeof(std::complex<T>) == sizeof(Kokkos::complex<T>)`, we should be able to make `memcpy` work by casting `Kokkos::complex<T>*` to `char*` and adjusting the `memcpy` length argument accordingly.

**Comment by mhoemmen** [comment]
@mhoemmen wrote:
> As long as `sizeof(std::complex<T>) == sizeof(Kokkos::complex<T>)`, we should be able to make `memcpy` work by casting `Kokkos::complex<T>*` to `char*` and adjusting the `memcpy` length argument accordingly.

Just to clarify: There are two ways to do this:

1. Change the implementation of `Kokkos::deep_copy` so it casts `T*` to `char*` before doing `memcpy`.
2. Change all code in Trilinos so that whenever Trilinos calls `Kokkos::deep_copy` on an `impl_scalar_type` `Kokkos::View`, Trilinos first turns the Views into nonowning Views of `char`.  This will ONLY work for contiguous rank-1 Views, though.

**Comment by crtrott** [comment]
Oh wait so the problem is inside of Kokkos::deep_copy?

**Comment by kddevin** [comment]
No, it is a memcpy in Tpetra

**Comment by mhoemmen** [comment]
@kddevin Christian says we can use `Kokkos::local_deep_copy` there in Tpetra, since it's a `memcpy` in a device function.  I'll try that change and see if it works.

**Comment by kddevin** [comment]
To be clearer:  it is a memcpy in which the input is Kokkos::Complex

**Comment by kddevin** [comment]
@mhoemmen The comments in the code say memcpy is fine with CUDA, and clearly it works.  But the gcc 8.3 compiler is pickier about the types.  
It seems artificial to have to avoid memcpy because of Kokkos::Complex.  I think we'll be bitten by this warning over and over.  (Of course, using Kokkos::Complex seems artificial too.)

**Comment by crtrott** [comment]
Kokkos::local_deep_copy is the abstraction inside of Kokkos which can deal with trivially vs non-trivially copyable. Also you will run into the same issues with sacado or stokhos types.

**Comment by mhoemmen** [comment]
@kddevin Christian points out that `Kokkos::local_deep_copy` is more general, in that it should be able to handle types that _really_ aren't trivially copyable.  It also has overloads for team parallelism.

**Comment by mhoemmen** [comment]
@kddevin Here's a PR for the Tpetra uses of memcpy: https://github.com/trilinos/Trilinos/pull/6394

**Comment by kddevin** [comment]
OK, for views, using local_deep_copy makes sense.  I don't understand creating Views just so we can use local_deep_copy, though.  I'm also concerned about relying on things in Kokkos::Experimental; what are Kokkos' plans for promoting local_deep_copy to the Kokkos namespace?

**Comment by nliber** [comment]
I may have a fix.

If we turn the ```volatile``` copy assignments into templated functions constrained with:

```
  template <class Complex,
            typename std::enable_if<std::is_same<Complex, complex>::value,
                                    int>::type = 0>

```
they are no longer copy assignment operators.  As long as they are exact matches, they will be called over trying to call a (now nonexistent) copy assignment operator, and all should be good.

I've got some more testing to do on this tomorrow...

**Comment by dhollman** [comment]
We need *slightly* more than that. Maintaining bug-for-bug compatibility requires that there also be a volatile copy conversion assignment for convertible (even narrowing) `RealType` (which doesn’t actually have to be floating point). I certainly hope that isn’t being used, but that’s what it did before.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by mhoemmen** [comment]
@dhollman wrote:
> Maintaining bug-for-bug compatibility requires that there also be a volatile copy conversion assignment for convertible (even narrowing) `RealType` (which doesn’t actually have to be floating point). I certainly hope that isn’t being used, but that’s what it did before.

I don't think we actually need that, at least as far as I know.  `Kokkos::complex` _only_ exists because 

  1. `std::complex` methods don't have `__device__` markings; and
  2. `std::complex` doesn't compile as the result of a `Kokkos::parallel_{reduce,scan}`, due to lack of `volatile` overloads.

I find it goofy that we need the `volatile` overloads -- I see that as a Kokkos implementation artifact that's unfortunately exposed to users.  Has that requirement gone away?  If so, then let's get rid of them.

**Comment by nliber** [comment]
@dhollman The only three I am looking to change are:

```
void operator=(const complex<RealType>& src) volatile noexcept { /* ... */ }
volatile complex& operator=(const volatile complex<RealType>& src) volatile noexcept { /* ... */ }
complex& operator=(const volatile complex<RealType>& src) noexcept { /* ... */ }
```

The other assignment operators (other than ```complex& operator=(const complex&) noexcept = default;```) are not copy assignment operators as far as the language is concerned, so should not impact making ```complex``` trivially copyable.

As I understand the rules in [class.copy.assign], a copy assignment operator for type X is:
 
Non-templated
Non-static
Exactly one parameter of type X&, where X is possibly cv-qualified

**Comment by nliber** [comment]
PR #2584

**Comment by masterleinad** [comment]
We decided to not include the fix in the release but will hotfix it for `Trilinos` promotion once it is ready.

**Comment by ndellingwood** [comment]
After promotion we can mark this InDevelop (it is not part of the release candidate branch)

**Comment by PhilMiller** [comment]
#5412 removes the `volatile` qualified constructors and all the other operators as long as `KOKKOS_ENABLE_DEPRECATED_CODE_4` is not set.

Sadly, that means this isn't close-able until Kokkos 5 ships, but it has been addressed.

**Comment by PhilMiller** [comment]
I misunderstood the earlier linked PR - it does make `Kokkos::Complex` trivially copyable, and there's unit test code to enforce this:
https://github.com/kokkos/kokkos/blame/develop/core/unit_test/TestComplex.hpp#L397
The fix has been present since the 3.1 release.
